tensordict/memmap.py

```diff
@@ -12,53 +12,42 @@
 from copy import copy, deepcopy
 from pathlib import Path
 from tempfile import _TemporaryFileWrapper
-from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+from typing import Any, Callable, Sequence
 
 import numpy as np
 import torch
 
 from tensordict.utils import (
     _getitem_batch_size,
-    DEVICE_TYPING,
-    INDEX_TYPING,
+    DeviceType,
+    IndexType,
+    NUMPY_TO_TORCH_DTYPE_DICT,
     prod,
-    torch_to_numpy_dtype_dict,
+    TORCH_TO_NUMPY_DTYPE_DICT,
 )
 
-MEMMAP_HANDLED_FN = {}
-
 __all__ = ["MemmapTensor", "set_transfer_ownership"]
 
+
 NoneType = type(None)
 EllipsisType = type(Ellipsis)
 
-numpy_to_torch_dtype_dict = {
-    "bool": torch.bool,
-    "uint8": torch.uint8,
-    "int8": torch.int8,
-    "int16": torch.int16,
-    "int32": torch.int32,
-    "int64": torch.int64,
-    "float16": torch.float16,
-    "float32": torch.float32,
-    "float64": torch.float64,
-    "complex64": torch.complex64,
-    "complex128": torch.complex128,
-}
-
-
-def implements_for_memmap(torch_function) -> Callable:
+
+MEMMAP_HANDLED_FN: dict[Callable, Callable] = {}
+
+
+def implements_for_memmap(torch_function: Callable) -> Callable:
     """Register a torch function override for ScalarTensor."""
 
     @functools.wraps(torch_function)
-    def decorator(func):
+    def decorator(func: Callable) -> Callable:
         MEMMAP_HANDLED_FN[torch_function] = func
         return func
 
     return decorator
 
 
-def to_numpy(tensor: Union[torch.Tensor, np.ndarray]) -> np.ndarray:
+def to_numpy(tensor: torch.Tensor | np.ndarray) -> np.ndarray:
```

Comment:
Can you elaborate on why we're moving away from union?

Reply:
They are semanticly identical. The `|` operator on classes syntax was introduced in Python 3.10 and can be enabled by `from __future__ import annotations` in Python 3.7. This is more clear and does not need unnecessary imports for `Union` and `Optional`.